futures-util/src/try_future/try_chain.rs

```diff
@@ -0,0 +1,66 @@
+use core::mem::PinMut;
+use futures_core::TryFuture;
+use futures_core::task::{Context, Poll};
+
+#[must_use = "futures do nothing unless polled"]
+#[derive(Debug)]
+crate enum TryChain<Fut1, Fut2, Data> {
```

Comment:
TryChain is a new mechanism so that `and_then` and `or_else` can share a single implementation.